R/intervals_support_funs.R

```diff
@@ -0,0 +1,295 @@
+    stop("The 'data' object must be a PKNCAdata object or a data frame.")
+  }
+  if (!is.character(target_impute)) {
+    stop("'target_impute' must be a character string.")
+  }
+  UseMethod("interval_add_impute")
+}
+
+#' @export
+interval_add_impute.PKNCAdata <- function(data, target_impute, after = Inf, target_params = NULL, target_groups = NULL, impute_column = NULL, allow_duplication = TRUE, new_rows_after_original = TRUE) {
+  if (is.null(impute_column) && !is.na(data$impute)) {
+    impute_column <- data$impute
+  }
+  data$intervals <- interval_add_impute(data$intervals, target_impute, after, target_params, target_groups, impute_column, allow_duplication, new_rows_after_original)
+  data
+}
+
+#' @export
+interval_add_impute.data.frame <- function(data, target_impute, after = Inf, target_params = NULL, target_groups = NULL, impute_column = NULL, allow_duplication = TRUE, new_rows_after_original = TRUE) {
+  
+  # Add an index column to preserve the original order
+  data <- dplyr::mutate(data, index = dplyr::row_number())
+  
+  # Get all parameter column names in the data frame
+  all_param_options <- setdiff(names(PKNCA::get.interval.cols()), c("start", "end"))
+  logical_cols <- names(which(colSums(data[sapply(data, is.logical)]) > 1))
+  param_cols <- intersect(logical_cols, all_param_options)
+  
+  # Handle target_params
+  if (is.null(target_params)) {
+    # Take all logical columns in data that are known parameters
+    target_params <- param_cols
+  } else {
+    # Check that all target_params are logical columns in data and known parameters
+    missing_params <- setdiff(target_params, logical_cols)
+    if (length(missing_params) > 0) {
+      stop("The following target_params are not interval columns and/or known PKNCA parameters: ", paste(missing_params, collapse = ", "))
+      target_params <- intersect(target_params, param_cols)
+    }
+  }
+  
+  # Determine the name of the impute column
+  impute_col <- if (!is.null(impute_column)) {
+    if (!impute_column %in% colnames(data)) {
+      stop("The 'data' object does not contain the specified impute column.")
+    }
+    impute_column
+  } else if ("impute" %in% colnames(data)) {
+    "impute"
+  } else {
+    data$impute <- NA_character_
+    "impute"
+  }
+  
+  # Identify the targeted intervals based on the groups
+  if (!is.null(target_groups)) {
+    target_intervals <- dplyr::inner_join(data, target_groups, by = names(target_groups))
+  } else {
+    target_intervals <- data
+  }
+  
+  # Identify the targeted intervals based on the parameters
+  target_intervals <- target_intervals %>%
+    dplyr::filter(rowSums(dplyr::across(dplyr::any_of(target_params), ~ . == TRUE)) > 0)
+  
+  # Add the imputation method to the targeted intervals
+  new_intervals_with_impute <- target_intervals %>%
+    dplyr::mutate(dplyr::across(dplyr::any_of(param_cols), ~FALSE)) %>%
+    dplyr::mutate(dplyr::across(dplyr::any_of(target_params), ~TRUE)) %>%
```

Comment:
This will cause a bug if not all `target_param` values are `TRUE` on a row because you would now calculate them all. For example, if `target_param = c("cmax", "tmax")` and `data = data.frame(start = 0, end = 24, cmax = c(TRUE, FALSE), tmax = TRUE)`, the output will have `cmax = TRUE` for both rows.

Please add a test to verify that this does not occur.